core/src/action.rs

```diff
@@ -0,0 +1,202 @@
+use agent::state::AgentState;
+use context::Context;
+use hash_table::entry::Entry;
+use holochain_dna::Dna;
+use instance::Observer;
+use nucleus::{state::NucleusState, EntrySubmission, FunctionCall, FunctionResult};
+use snowflake;
+use std::{
+    hash::{Hash, Hasher},
+    sync::{mpsc::Sender, Arc},
+};
+
+#[derive(Clone, Debug)]
+// @TODO what is wrapper for?
+// https://github.com/holochain/holochain-rust/issues/192
+pub struct ActionWrapper {
+    action: Action,
+    id: snowflake::ProcessUniqueId,
+}
+
+impl ActionWrapper {
+    /// constructor from &Action
+    /// internal snowflake ID is automatically set
+    pub fn new(a: &Action) -> Self {
+        ActionWrapper {
+            action: a.clone(),
+            // auto generate id
+            id: snowflake::ProcessUniqueId::new(),
+        }
+    }
+
+    /// read only access to action
+    pub fn action(&self) -> Action {
+        self.action.clone()
+    }
+}
+
+impl PartialEq for ActionWrapper {
+    fn eq(&self, other: &ActionWrapper) -> bool {
+        self.id == other.id
+    }
+}
+
+impl Eq for ActionWrapper {}
+
+impl Hash for ActionWrapper {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.id.hash(state);
+    }
+}
+
+#[derive(PartialEq, Clone, Hash, Debug)]
+pub struct Action {
+    signal: Signal,
+    id: snowflake::ProcessUniqueId,
+}
+
+impl Action {
+    /// constructor from &Signal
+    /// snowflake ID is auto generated
+    pub fn new(signal: &Signal) -> Action {
+        Action {
+            signal: signal.clone(),
+            // auto generate id
+            id: snowflake::ProcessUniqueId::new(),
+        }
+    }
+
+    /// read only access to the internal Signal
+    pub fn signal(&self) -> Signal {
+        self.signal.clone()
+    }
+}
+
+impl Eq for Action {}
+
+#[derive(Clone, PartialEq, Hash, Debug)]
+pub enum Signal {
```

Comment:
@ddd-mtl it has not been renamed signal, there were three (four?) structural levels before this branch, one was just implicit so i made it explicit and gave it a name

there's a followup to do what you're suggesting https://github.com/holochain/holochain-rust/issues/192

Reply:
the purpose of the changes i made other than naming things is consolidation. i'm working towards:

- a flatter structure overall (e.g. something like what you suggested) so less `match`ing
- a single snowflake ID per action dispatch
- a single standard representation for key/values for action/result pairs across all state slices
- a single enum (rather than struct) that defines which types go where

happy to rename `Signal` if its confusing, there aren't many good words in this space though